### ISSUE TITLE
Change favicon (docker wiki?)

### ISSUE THREAD

**Issue by niawag** [bpatrik/pigallery2#325]
Hi and thanks a lot for this great app, it's working great !

I'm not submitting a bug but, as I wanted to change the site favicon, I've spent a bit of time to find where the icon was stored in the  container and how to replace it.
I found the icon with : `docker export %your container id%| tar tf - | grep icon.png`
Then, I replaced it using docker-compose volumes:
```
    volumes:
      - "/YOUR OWN PATH TO/config:/app/data/config"
      - "/YOUR OWN PATH TO/db-data:/app/data/db" #to make db persistent (instead of "db-data" only)
      - "/YOUR OWN PATH TO/photo:/app/data/images"
      - "/YOUR OWN PATH TO/tmp:/app/data/tmp"
      #to change favicon for english and french locale
      - "/storage/dockers/pigallery2/icon.png:/app/dist/en/assets/icon.png"
      - "/storage/dockers/pigallery2/icon.png:/app/dist/fr/assets/icon.png"
      #to change icon next to title for english and french locale 
      - "/storage/dockers/pigallery2/icon.png:/app/dist/en/assets/icon_inv.png"
      - "/storage/dockers/pigallery2/icon.png:/app/dist/fr/assets/icon_inv.png"
```
I hope this could help others trying to achieve the same thing! If you think it's usefull I'd be glad to create a wiki section.
Let me know if there is any problem related to the changes I made ! (icon or db-data related)

**Comment by bpatrik** [comment]
Feels a bit of a hack to be honest. 

I would not mind making it possible to change the icon from settings. Let me know if you would like to take a look, i can give some pointers. 

Also `"db-data:/app/data/db"` should already persist the db in a volume. Even if you delete a container that should remain intact.

**Comment by niawag** [comment]
Yes, hacky it is! But it's also a clever way to change ressources inside a container (I guess I like this kind of hack...)

I'd love to help with making the icon change as a setting option and I'll need some guidance/pointers to have a quick start.

About db-data, I kept looking into it and it seems to be a problem related to the docker-compose library used in Portainer. It forces me to remove the db-data volume each time I change a setting in the docker yaml, which I did a few time for my hack to work.

**Comment by bpatrik** [comment]
Great!

What you can do is to create a new icon url (or path) in the config:
https://github.com/bpatrik/pigallery2/blob/41975f72a18f985cae27b1bdb6b31c7379b0d87c/src/common/config/public/ClientConfig.ts#L174-L180
(same for the icon_inv)

Then you can do multiple things:
One option would be just to override the `/assets/icon.png` middleware here to server the configured icon, instead of the one under `/assets/icon.png`:
https://github.com/bpatrik/pigallery2/blob/41975f72a18f985cae27b1bdb6b31c7379b0d87c/src/backend/routes/PublicRouter.ts#L132-L139
This way you do not need to change the front-end.

An other would be to change all link in the UI to point to what it was given in the config:
https://github.com/bpatrik/pigallery2/search?q=icon.png
But this is more hassle and I think the previous option would be better.

----------------------------------
optionally you can also put the setting on the UI:

For that you would need to:
1. map to UI config:
https://github.com/bpatrik/pigallery2/blob/41975f72a18f985cae27b1bdb6b31c7379b0d87c/src/frontend/app/ui/settings/basic/basic.settings.component.ts#L26-L34
2. show on the UI:
https://github.com/bpatrik/pigallery2/blob/41975f72a18f985cae27b1bdb6b31c7379b0d87c/src/frontend/app/ui/settings/basic/basic.settings.component.html#L10-L27
3. Implement the backend part on config save (do the mapping backwards):
https://github.com/bpatrik/pigallery2/blob/41975f72a18f985cae27b1bdb6b31c7379b0d87c/src/backend/middlewares/admin/SettingsMWs.ts#L391-L398

**Comment by niawag** [comment]
Thanks a lot for your help, I think I should be able to handle most of it.

There is just this part where I'm not sure about how to do it: 

> Then you can do multiple things:
> One option would be just to override the `/assets/icon.png` middleware here to server the configured icon, instead of the one under `/assets/icon.png`:
> https://github.com/bpatrik/pigallery2/blob/41975f72a18f985cae27b1bdb6b31c7379b0d87c/src/backend/routes/PublicRouter.ts#L132-L139

Should I add an if statement in both app.get functions ? Somtehing like 
if ('file' == 'icon.png') {//change file to Config.Client.icon}
else //normal behaviour
OR
edit renderFile and check the file to render, if it's icon.png change file to Config.Client.icon

How can we add our own icons? If it's a URL it's simple but if someone wants to serve the icon locally, can we add an icon in config and serve it from there or should we make the assets folder as a docker volume?

**Comment by bpatrik** [comment]
you can also do something like:
```

    app.get('/assets/icon.png',
      setLocale,
 (req: Request, res: Response, next: NextFunction)=>{
//maybe some checks here? OR when the config is set ( see Diagnostic class)
 res.sendFile(<path to the icon>);
});
```

**Comment by bpatrik** [comment]
If the icon is an URL that is actually the more tricky one, because you will need to change the fontend where the icon is referenced. 

If its a local file, you can use the code above.

What would also be a nice solution is:

1) if there is an icon file in eg the /config folder as you also suggested, then the app would use that icon, otherwise the default. 
2) Create an upload filed in the configs that would save the icon to that config folder. 

* With this solution you would not need to create a new config field.

**Comment by niawag** [comment]
Hey, I like the KISS solution with an icon file in the config folder, I'll try it as soon as possible! 
I may start with step 1 that will bring the major functionnality (changing icons) then step 2.

**Comment by niawag** [comment]
Here is a first attempt at a pull request: https://github.com/bpatrik/pigallery2/pull/331
